### PR TITLE
[RFC] Thread cleanup

### PR DESCRIPTION
**Pull request by jenswi-linaro**
This is a prerequisite for the SPCI patches as they look now. These patches makes sense on their own. So in order to avoid unnecessary tricky rebasing it would be nice to get these through a bit earlier.

By the way, the SPCI patches are still evolving and will continue to do so for some time.

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
Force pushed a fix for the virtualization compile error (missing #include).

**Comment by jenswi-linaro** [comment]
Two platforms stands out here:
- plat-stm overriding `tee_entry_std()`
- plat-zynq7k overriding `tee_entry_fast()`

**Comment by jenswi-linaro** [comment]
Rebasing on master to resolve merge conflict

File: core/arch/arm/kernel/sub.mk [hunk 1]
```diff
@@ -28,6 +28,9 @@ srcs-$(CFG_PL310) += tee_l2cc_mutex.c
 srcs-$(CFG_ARM32_core) += thread_a32.S
 srcs-$(CFG_ARM64_core) += thread_a64.S
 srcs-y += thread.c
+srcs-y += thread_osmci.c
+srcs-$(CFG_ARM32_core) += thread_osmci_a32.S
+srcs-$(CFG_ARM64_core) += thread_osmci_a64.S
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/kernel/sub.mk]: Fixed in the commit.

File: core/arch/arm/kernel/thread_private.h [hunk 2]
```diff
@@ -147,8 +147,16 @@ void thread_excp_vect_workaround(void);
 void thread_excp_vect_workaround_a15(void);
 void thread_excp_vect_end(void);
 
-/* Handles a stdcall, r0-r7 holds the parameters */
-void thread_std_smc_entry(void);
+/*
+ * Assembly function as the first funciton in a thread.  Handles a stdcall,
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/kernel/thread_private.h]: Fixed in the commit.

File: core/arch/arm/kernel/thread_osmci_a32.S [hunk 3]
```diff
@@ -230,8 +230,8 @@ UNWIND(	.save	{r0})
 	 * function was originally entered.
 	 */
 	pop	{r12}			/* Get pointer to rv[] */
-	stm	r12, {r0-r5}		/* Store r0-r5 into rv[] */
-	pop	{r4-r5, pc}
+	stm	r12, {r0-r3}		/* Store r0-r3 into rv[] */
+	pop	{r4-r5, pc} /* TODO r4-r5 */
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/kernel/thread_osmci_a32.S]: Pushed a "[review]" commit fixing this.

**Comment by jenswi-linaro** [comment]
Rebased on master to try to fix IBART build error

File: core/arch/arm/kernel/thread_osmci_a32.S (hunk 3 above)
[the remark this answers is not in the document]
Comment (jforissier) [reply core/arch/arm/kernel/thread_osmci_a32.S]: Fine with me

**Comment by jenswi-linaro** [comment]
Squashed and tag applied

**Pull request merged**